File: routeMap/app/src/main/java/ru/android/zheka/vm/PanelHomeVM.java
package ru.android.zheka.vm;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Intent;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import java.io.InputStream;
import java.util.Scanner;

import io.reactivex.functions.Consumer;
import ru.android.zheka.coreUI.ButtonHandler;
import ru.android.zheka.coreUI.IActivity;
import ru.android.zheka.gmapexample1.AddressActivity;
import ru.android.zheka.gmapexample1.EditActivity;
import ru.android.zheka.gmapexample1.GeoPositionActivity;
import ru.android.zheka.gmapexample1.LatLngActivity;
import ru.android.zheka.gmapexample1.PointToTraceActivity;
import ru.android.zheka.gmapexample1.PositionInterceptor;
import ru.android.zheka.gmapexample1.SettingsActivity;
import ru.android.zheka.gmapexample1.edit.EditModel;
import ru.android.zheka.model.IHomeModel;

import static ru.android.zheka.gmapexample1.R.id;
import static ru.android.zheka.gmapexample1.R.layout;
import static ru.android.zheka.gmapexample1.R.raw;
import static ru.android.zheka.gmapexample1.R.string;

public class PanelHomeVM implements IPanelHomeVM {

    IActivity view;
    IHomeModel model;

    public PanelHomeVM(IActivity view, IHomeModel model) {
        this.view = view;
        this.model = model;
    }

    @Override
    public void info() {
        Activity activity = view.getActivity ();
        LayoutInflater inflater = LayoutInflater.from (activity);
        View view = inflater.inflate (layout.scrolable_dialog, null);
        TextView tv = view.findViewById (id.textmsg);//new TextView (this);
        InputStream is = this.view.getActivity ().getResources ().openRawResource (raw.info);
        Scanner scanner = new Scanner (is);
        scanner.useDelimiter ("\n");
        StringBuilder sb = new StringBuilder ();
        while (scanner.hasNextLine ())
            sb.append (scanner.nextLine ());
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            tv.setText (Html.fromHtml (sb.toString (), Html.FROM_HTML_MODE_LEGACY));
        } else
            tv.setText (Html.fromHtml (sb.toString ()));
        new AlertDialog.Builder (activity)
                .setView (view)
                .setTitle (activity.getResources ().getString (string.home_info_btn))
                .setCancelable (true)
                .create ()
                .show ();
    }

    @Override
    public void address() {

        Intent intent = view.getActivity ().getIntent ();
        intent.setAction (Intent.ACTION_VIEW);
        intent.setClass (view.getActivity (), AddressActivity.class);
        view.getActivity ().startActivity (intent);
        view.getActivity ().finish ();
    }

    @Override
    public void geo() {
        PositionInterceptor position = new PositionInterceptor (view.getActivity ());
        Intent intent = position.updatePosition ();
        intent.setClass (view.getContext (), GeoPositionActivity.class);
        intent.setAction (Intent.ACTION_VIEW);
        //explicit activity
        view.getActivity ().startActivity (intent);
        view.getActivity ().finish ();
    }

    @Override
    public void editPoints() {
        editItem ("Point", string.points_column_name, string.points_column_name1);
    }

    @Override
    public void editTraces() {
        editItem ("Trace", string.traces_column_name, string.traces_column_name1);
    }

    public int editItem(String item, int nameId, int name1Id) {
        Intent intent = view.getActivity ().getIntent ();
        EditModel model = new EditModel ();
        model.clsName = item;
        model.clsPkg = "ru.android.zheka.db";
        model.name1Id = name1Id;
        model.nameId = nameId;
        intent.putExtra (EditActivity.EDIT_MODEL, model);
        intent.setAction (Intent.ACTION_VIEW);
        intent.setClass (view.getContext (), EditActivity.class);
        view.getActivity ().startActivity (intent);
        view.getActivity ().finish ();
        return 0;
    }

    @Override
    public void pointNavigate() {
        Intent intent = view.getActivity ().getIntent ();
        intent.setClass (view.getContext (), LatLngActivity.class);
        intent.setAction (Intent.ACTION_VIEW);
        view.getActivity ().startActivity (intent);
        view.getActivity ().finish ();
    }

    @Override
    public void createTrace() {
        Intent intent = view.getActivity ().getIntent ();
        intent.setClass (view.getActivity (), PointToTraceActivity.class);
        intent.setAction (Intent.ACTION_VIEW);
        view.getActivity ().startActivity (intent);
        view.getActivity ().finish ();
    }

    @Override
    public void settingsAction() {
        Intent intent = view.getActivity ().getIntent ();
        intent.setAction (Intent.ACTION_VIEW);
        intent.setClass (view.getContext (), SettingsActivity.class);
        view.getActivity ().startActivity (intent);
        view.getActivity ().finish ();
    }

    private ButtonHandler getButton(Consumer <Boolean> consumer, int nameId) {
        return new ButtonHandler (consumer
                , nameId
                , view);
    }

    @Override
    public void onResume() {
        model.getStartButton ().set (getButton (a -> settingsAction (), string.home_settings_btn));
        model.getStopButton ().set (getButton (a -> address (), string.home_address_btn));
        model.getNextButton ().set (getButton (a -> pointNavigate (), string.home_points_btn));
        model.getStartButton1 ().set (getButton (a -> editTraces (), string.home_editTrace_btn));
        model.getStopButton1 ().set (getButton (a -> editPoints (), string.home_editPoint_btn));
        model.getNextButton1 ().set (getButton (a -> createTrace (), string.home_toTrace_btn));
        model.getStartButton2 ().set (getButton (a -> geo (), string.home_geo_btn));
        model.getStopButton2 ().set (getButton (a -> info (), string.home_info_btn));
        model.inputVisible ().set (View.GONE);
    }

    @Override
    public void onDestroy() {
    }

    @Override
    public IHomeModel model() {
        return model;
    }
}

File: routeMap/app/src/main/java/ru/android/zheka/vm/PanelHomeVM.kt
package ru.android.zheka.vm

import android.app.AlertDialog
import android.content.Intent
import android.os.Build
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import androidx.fragment.app.Fragment
import io.reactivex.functions.Consumer
import ru.android.zheka.coreUI.ButtonHandler
import ru.android.zheka.coreUI.IActivity
import ru.android.zheka.coreUI.IPanelModel
import ru.android.zheka.db.DbFunctions
import ru.android.zheka.db.Point
import ru.android.zheka.fragment.*
import ru.android.zheka.gmapexample1.GeoPositionActivity
import ru.android.zheka.gmapexample1.PositionInterceptor
import ru.android.zheka.gmapexample1.R.*
import ru.android.zheka.model.HomeModel
import ru.android.zheka.model.IHomeModel
import java.util.*

class PanelHomeVM(val view: IActivity, val model: IPanelModel,
                  val edit: IEdit, val editTraces: IEditTraces,
                  val trace: ITrace, val enterPoint: IEnterPoint
) : IPanelHomeVM {
    var fragment: Fragment?

    init {
        fragment = view.manager?.findFragmentById(id.latLngFragment)
    }

    override fun settings() {
        removeAndSwitchFragmentFromContainer(id.latLngFragment, Settings())
    }

    private fun removeAndSwitchFragmentFromContainer(id: Int, fragment_: Fragment) {
        if (fragment != null) {
            view.removeFragment(fragment!!)
        }
        fragment = fragment_
        view.switchToFragment(id, fragment!!)
    }

    override fun info() {
        val activity = view.activity
        val inflater = LayoutInflater.from(activity)
        val view = inflater.inflate(layout.scrolable_dialog, null)
        val tv = view.findViewById<TextView>(id.textmsg) //new TextView (this);
        val `is` = this.view.activity.resources.openRawResource(raw.info)
        val scanner = Scanner(`is`)
        scanner.useDelimiter("\n")
        val sb = StringBuilder()
        while (scanner.hasNextLine()) sb.append(scanner.nextLine())
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            tv.text = Html.fromHtml(sb.toString(), Html.FROM_HTML_MODE_LEGACY)
        } else tv.text = Html.fromHtml(sb.toString())
        AlertDialog.Builder(activity)
                .setView(view)
                .setTitle(activity.resources.getString(string.home_info_btn))
                .setCancelable(true)
                .create()
                .show()
    }

    override fun enterPoint() {
        removeAndSwitchFragmentFromContainer(id.latLngFragment, enterPoint as EnterPoint)
    }

    override fun geo() {
        val position = PositionInterceptor(view.activity)
        val intent = position.updatePosition()
        intent.setClass(view.context, GeoPositionActivity::class.java)
        intent.action = Intent.ACTION_VIEW
        //explicit activity
        view.activity.startActivity(intent)
        view.activity.finish()
    }

    override fun editPoints() {
//        editItem("Point", string.points_column_name, string.points_column_name1)
        removeAndSwitchFragmentFromContainer(id.latLngFragment, edit as Edit)
    }

    override fun editTraces() {
//        editItem("Trace", string.traces_column_name, string.traces_column_name1)
        removeAndSwitchFragmentFromContainer(id.latLngFragment, editTraces as EditTraces)
    }

    override fun pointNavigate() {
        removeAndSwitchFragmentFromContainer(id.latLngFragment, LatLng())
    }

    override fun createTrace() {
        removeAndSwitchFragmentFromContainer(id.latLngFragment, trace as Trace)
    }

    private fun getButton(consumer: Consumer<Boolean>, nameId: Int): ButtonHandler {
        return ButtonHandler(consumer
                , nameId
                , view)
    }

    override fun onResume() {
        model.inputVisible().set(View.GONE)
        model.startButton.set(getButton(Consumer { settings() }, string.home_settings_btn))
        model.stopButton.set(getButton(Consumer { enterPoint() }, string.home_address_btn))
        if (isPoints()) {
            model.nextButton.set(getButton(Consumer { pointNavigate() }, string.home_points_btn))
            model.stopButton1.set(getButton(Consumer { editPoints() }, string.home_editPoint_btn))
            model.nextButton1.set(getButton(Consumer { createTrace() }, string.home_toTrace_btn))
        }
        model.startButton1.set(getButton(Consumer { editTraces() }, string.home_editTrace_btn))
        model.startButton2.set(getButton(Consumer { geo() }, string.home_geo_btn))
        model.stopButton2.set(getButton(Consumer { info() }, string.home_info_btn))
    }

    private fun isPoints(): Boolean {
        val list = DbFunctions.getTablesByModel(Point::class.java)
        return list != null && list.isNotEmpty()
    }

    override fun onDestroy() {}
    override fun model(): IHomeModel {
        return model as HomeModel
    }
}